### ISSUE TITLE
Invalid order in SASL authentication call resulting wrong message which trigger connection reset from the server

### ISSUE THREAD

**Issue by yulrizka** [segmentio/kafka-go#542]
**Describe the bug**
using Writer struct with SASL authentication resulted sending the wrong message and made the
server terminate the connection

**Kafka Version**
Azure EventHub 

**To Reproduce**
Create a writer with SASL and TLS using the struct directly
```
	mechanism := plain.Mechanism{
		Username: "xxx",
		Password: "xxx",
	}


	transport := &kafka.Transport{
		Dial: (&net.Dialer{
			Timeout:   10 * time.Second,
			DualStack: true,
		}).DialContext,
		SASL: mechanism,
		TLS: &tls.Config{
			InsecureSkipVerify: true,
		},
	}
	w := kafka.Writer{
		Addr:        kafka.TCP("someurl.windows.net:9093"),
		Topic:       ActivityLogTopic,
		Balancer:    &kafka.LeastBytes{},
		ErrorLogger: kafka.LoggerFunc(log.Errorf),
		Logger:      kafka.LoggerFunc(log.Infof),
		Async:       false,
		Transport:   transport,
	}
```

**Expected behavior**
Not throwing any error

**Additional context**
I did some debugging and found that SASL handshake was performed before getting all available kafka api version.

I made a PR https://github.com/segmentio/kafka-go/pull/541/files

More specifically, in this code, the order should be reversed (also aligned with the [protocol](https://kafka.apache.org/protocol#sasl_handshake))
https://github.com/segmentio/kafka-go/blob/cb3dd797e4da19c25b2aa7b976986df98eb822c4/transport.go#L1138-L1158

Since the order was reverse, the handshake was done with emtpy `c.versions` which resulted in wrong api `0` instead of `1`
https://github.com/segmentio/kafka-go/blob/cb3dd797e4da19c25b2aa7b976986df98eb822c4/protocol/conn.go#L81-L91

**Comment by achille-roussel** [comment]
Thanks for the detailed bug report @yulrizka!

I merged your change and tagged a release for v0.4.6 containing it 👍

**Comment by yulrizka** [comment]
@achille-roussel do you mind tagging this PR with `hacktoberfest-accepted` @achille-roussel? I'm participating in [Hacktoberfest](https://hacktoberfest.digitalocean.com/) :)
